File: app/src/main/java/pl/mosenko/sunnypodlaskie/SettingsFragment.java
package pl.mosenko.sunnypodlaskie;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceFragmentCompat;

import pl.mosenko.sunnypodlaskie.util.PreferenceWeatherUtil;
import pl.mosenko.sunnypodlaskie.util.WeatherDataAlarmSyncUtil;
import pl.mosenko.sunnypodlaskie.util.WeatherDataJobSyncUtils;

/**
 * Created by syk on 23.05.17.
 */

public class SettingsFragment extends PreferenceFragmentCompat implements SharedPreferences.OnSharedPreferenceChangeListener {

    private SharedPreferences mSharedPreferences;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.pref_general);
        mSharedPreferences = getPreferenceScreen().getSharedPreferences();

    }

    @Override
    public void onStart() {
        super.onStart();
        mSharedPreferences.registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onStop() {
        super.onStop();
        mSharedPreferences.registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        final Context context = getContext();
        boolean changedSyncTime = key.equals(getString(R.string.pref_sync_time_key));

        if (changedSyncTime) {
            addSyncTimePreferenceSummary(sharedPreferences, key);
            WeatherDataAlarmSyncUtil.startAlarm(context);
        }
    }

    private void addSyncTimePreferenceSummary(SharedPreferences sharedPreferences, String key) {
        Preference preference = findPreference(key);
        preference.setSummary(sharedPreferences.getString(key, ""));
    }
}

File: app/src/main/java/pl/mosenko/sunnypodlaskie/ui/setting/SettingsFragment.kt
package pl.mosenko.sunnypodlaskie.ui.setting

import android.os.Bundle
import android.view.View
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.google.android.material.snackbar.Snackbar
import org.koin.androidx.viewmodel.ext.android.viewModel
import pl.mosenko.sunnypodlaskie.R

/**
 * Created by syk on 23.05.17.
 */
class SettingsFragment : PreferenceFragmentCompat() {

    val viewModel: SettingsViewModel by viewModel()

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.pref_general)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        configureSyncTimePreference()
    }

    override fun onStart() {
        super.onStart()
        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(viewModel)
    }

    override fun onStop() {
        super.onStop()
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(viewModel)
    }

    private fun configureSyncTimePreference() {
        val preference: Preference = findPreference(getString(R.string.pref_sync_time_key))!!
        preference.setOnPreferenceChangeListener { _, newValue ->
            preference.summary = newValue.toString()
            val success: Boolean = viewModel.validateSyncTime(newValue.toString())
            if (!success) {
                Snackbar.make(
                    requireView(),
                    R.string.bad_sync_time_format_message,
                    Snackbar.LENGTH_SHORT
                ).show()
            }
            success
        }
        viewModel.syncTime.observe(viewLifecycleOwner, {
            preference.summary = it
        })
    }
}
